daemon/hooks/hook_plan.go

```diff
@@ -0,0 +1,249 @@
+package hooks
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/zrepl/zrepl/zfs"
+)
+
+type HookReport struct {
+	Command                      string
+	Args                         []string // currently always empty
+	Env                          []string
+	Edge                         HookEdge
+	RunTime                      time.Duration
+	Err                          error
+	CapturedStdoutStderrCombined []byte
+}
+
+func (e *HookReport) Error() string {
+	if e.Err == nil {
+		return ""
+	}
+	cmdLine := strings.Join(append([]string{e.Command}, e.Args...), " ")
+	return fmt.Sprintf("%s-hook failed on command \"%s\" with error: %s", e.Edge, cmdLine, e.Err)
+}
+
+func (e *HookReport) HadError() bool {
+	return e.Err != nil
+}
+
+type hookDependencySet map[*hookPlanEntry]struct{}
+
+func (d *hookDependencySet) Add(e *hookPlanEntry) {
+	for key := range *d {
+		if e.Order == key.Order {
+			return
+		}
+	}
+
+	(*d)[e] = struct{}{}
+}
+
+func (d *hookDependencySet) OrdersAsSlice() (ret []int) {
+	for key := range *d {
+		ret = append(ret, key.Order)
+	}
+	sort.Ints(ret)
+	return
+}
+
+type hookPlanEntry struct {
+	Order        int
+	Edge         HookEdge
+	Hook         Hook
+	ErrIsFatal   bool
+	Dependencies hookDependencySet
+}
+
+type HookJobPlan struct {
+	mtx     sync.RWMutex
+	context context.Context
+	steps   []hookPlanEntry // protected by mtx
+	phase   HookPhase
+	env     HookEnv
+}
+
+func NewHookJobPlan(ctx context.Context, hooks *HookList, phase HookPhase, fs *zfs.DatasetPath, cb HookJobCallback, extra HookEnv) (*HookJobPlan, error) {
+	l := getLogger(ctx)
+
+	plan := &HookJobPlan{
+		context: ctx,
+		phase:   phase,
+		env:     extra,
+	}
+
+	for _, h := range *hooks {
+		//l := l.WithField("hook", h)
+
+		passFilesystem, err := h.Filesystems().Filter(fs)
+		if err != nil {
+			return nil, fmt.Errorf("unexpected filter error: %s", err)
+		}
+		if passFilesystem {
+			/*
+			   Iterate over each edge in the bitfield,
+			   creating multiple entries in the plan if
+			   the hook has multiple edges.
+			*/
+			//l.Debug(fmt.Sprintf("adding hook to edge %s", edge.String()))
```

Comment:
or leave it in and enable this debug stmt, if you think it will need / aid debugging.
consider introducing a `func debug` enabled by an env variable as found in other packages (e.g. rpc)